src/cargo/ops/cargo_doc.rs

```diff
@@ -117,6 +126,80 @@ pub fn doc(ws: &Workspace<'_>, options: &DocOptions) -> CargoResult<()> {
     Ok(())
 }
 
+fn merge_cross_crate_info(ws: &Workspace<'_>, compilation: &Compilation<'_>) -> CargoResult<()> {
+    let Some(fingerprints) = compilation.rustdoc_fingerprints.as_ref() else {
+        return Ok(());
+    };
+
+    let now = std::time::Instant::now();
+    for (kind, fingerprint) in fingerprints.iter() {
+        let (target_name, build_dir, artifact_dir) = match kind {
+            CompileKind::Host => ("host", ws.build_dir(), ws.target_dir()),
+            CompileKind::Target(t) => {
+                let name = t.short_name();
+                let build_dir = ws.build_dir().join(name);
+                let artifact_dir = ws.target_dir().join(name);
+                (name, build_dir, artifact_dir)
+            }
+        };
+
+        // rustdoc needs to read doc parts files from build dir
+        build_dir.open_ro_shared_create(".cargo-lock", ws.gctx(), "build directory")?;
```

Comment:
non-blocker: technically there is a race condition between the build and this but the window is small and likely the users fault if they have a  `cargo clean` running in parallel, blocked on the build.

This would be solved by us moving the merge into the build graph